### DISCUSSION TITLE
How do I update 60K+ rows efficiently?

### DISCUSSION THREAD

**Discussion by Krabba** [prisma/prisma#14708] [Q&A]
Hi,

Sorry if initially i'm not providing enough information.

I'm trying to update 62993 rows in my **PostgreSQL database** with different data, `updateMany` is not an option since there are no common denominators but I can still use it to update individual values one at a time.

I've gotten `$executeRaw` to perform ONE update statement (where I nested tons and tons of `CASE` + `WHEN` + `THEN` statements following the `SET` statement), I stored that entire statement in a string (by concatenating logic in `for` loops from JavaScript), which came out to be about 1.2 million characters long.
Needless to say, it worked on a few rows, but this solution didn't scale and ended up freezing my terminal/database.

So how am I meant to update many rows, with no common denominators in the most efficient way possible?
I'm trying not to invoke 62993 `prisma.table.update()` or `prisma.table.updateMany()` calls.

**Comment by austincrim** [comment]
Hey @Krabba 👋,

Is this a kind of one-off update or something you expect to happen regularly?

If it is one-off, I would run the query(ies) in some kind of SQL workbench, not through Prisma.

**Reply by Krabba** [reply]
Hello @austincrim
Thank you for your reply. 😊

These fields in the database will be updated on a daily basis, so once per day through a cronjob.
I chose not to create new rows everyday, since that will quickly add up to a lot of data instead of just incrementing a counter on my current data which is about 63K rows, the counter is different for each row so I couldn't just say "give +1 to every row" which is where `updateMany` would be perfect.

If the total count in four different rows came out to be this today:
- Row A: 250
- Row B: 324
- Row C: 2
- Row D: 74

I can **NOT** do something like this:
```sql
UPDATE "public"."table"
SET "count" = 250,
SET "count" = 324,
SET "count" = 2,
SET "count" = 74
WHERE "id" = 1, "id" = 2, "id" = 3, "id" = 4
```

Which is where the `CASE` & `WHEN` & `THEN` statements come in handy.
Because then I can do this:

```sql
UPDATE "public"."table"
SET "count" = CASE "id"
    WHEN 1 THEN 250
    WHEN 2 THEN 324
    WHEN 3 THEN 2
    WHEN 4 THEN 74
    ELSE "count" END
```

That would allow me to do one database call which is much better than 63K database calls, but in my example I need to nest another set of `CASE` & `WHEN` & `THEN` inside the already "complex" `CASE` statement for the update, which for 63K rows, comes out to be a string of 1.2 million characters i'm sending to `$queryRaw` / `$queryRawUnsafe` / `$executeRaw` / `$executeRawUnsafe`.

Obviously it's too much to handle for the terminal and/or database, so things are freezing up, but could work for smaller batches perhaps, but even then I am not comfortable with the amount of database calls.

Just wondering if there is any clever idea out there to avoid that many database calls for a daily routine.

**Reply by austincrim** [reply]
Hm, I'm not sure we have a ton of options then. Your best bet is probably to try to batch them up into multiple queries. It might take some experimentation to find an optimal batch size, but that's what I would do.

**Reply by vladi-stevanovic** [reply]
👋 Hello @Krabba I'm going through pending GitHub Discussions to check if you need any help. Is this issue still open?
Please let us know if we can investigate it further or if you've already solved it, we'd love to know if you ended up following Austin's recommendation.

It would be super useful for other users facing this problem! 🙌

**Reply by Krabba** [reply]
Hey @vladi-stevanovic !

I am still trying to find the most optimal way to handle this use case.

But, with that said.

I did find that doing approximately 85 database calls was an okay number.

Each call consists of around 800-1300 rows.

This was fine to my local or prod environment Docker container (postgres).

But I wanted to expand, PlanetScale does not seem to accept that many updates. So that is why I am thinking again. 👏

**Reply by pantharshit00** [reply]
Hello, 

I think you should open a feature request about this. I wasn't able to find a matching feature request for SQL CASE...THEN support so I will suggest you to open one. 

This one might be related: https://github.com/prisma/prisma/issues/5560

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
